torch_xla/csrc/ops/gather.cpp

```diff
@@ -7,31 +7,44 @@
 #include "torch_xla/csrc/ops/infer_output_shape.h"
 
 namespace torch_xla {
+namespace {
+
+xla::Shape NodeOutputShape(const torch::lazy::Value& input,
```

Comment:
hmm, will `IsSparseGather` change the output shape? Why do we need a `NodeOutputShape` here?

Reply:
It does, since dense gather actually uses reduce instead of gather.

Reply:
FYI https://pytorch.org/docs/stable/generated/torch.gather.html. The output tensor of `torch.gather` has the same shape as `index`.

An issue I found with XLA dense gather is that it does not allow `index.size(d) < input.size(d)` for dimension `d != dim` so that the following code does not work with dense gather. Perhaps this type of input is not typical since nobody has complained.
```
a = torch.randint(0, 10, size=(8, 8), dtype=torch.int64, device=device)
b = torch.ones(10, 10).to(device)
c = torch.gather(b, dim=1, index=a)
```

Reply:
I see, @ymwangg let me check with XLA folks about this before landing, and see if they could modify the implementation or the check(s).

Reply:
@ymwangg I think it's a bug and will file a separate bug to track internally. Thanks!